quasiquotes3/shared/src/main/scala/scala/meta/quasiquotes/Api.scala

```diff
@@ -0,0 +1,196 @@
+      parse(input, dialect)
+    }.orElse{
+      val parse = implicitly[Parse[Pat]]
+      parse(input, dialect)
+    } match {
+      case x: scala.meta.parsers.Parsed.Success[_] => x.tree
+      case x: scala.meta.parsers.Parsed.Error => throw x.details
+    }
+  }
+  object XTensionQuasiquoteInit {
+    private[meta] def parse(input: scala.meta.inputs.Input, dialect: scala.meta.Dialect) = {
+      val parse = implicitly[Parse[Init]]
+      parse(input, dialect)
+    } match {
+      case x: scala.meta.parsers.Parsed.Success[_] => x.tree
+      case x: scala.meta.parsers.Parsed.Error => throw x.details
+    }
+  }
+  object XTensionQuasiquoteSelf {
+    private[meta] def parse(input: scala.meta.inputs.Input, dialect: scala.meta.Dialect) = {
+      val parse = implicitly[Parse[Self]]
+      parse(input, dialect)
+    } match {
+      case x: scala.meta.parsers.Parsed.Success[_] => x.tree
+      case x: scala.meta.parsers.Parsed.Error => throw x.details
+    }
+  }
+  object XTensionQuasiquoteTemplate {
+    private[meta] def parse(input: scala.meta.inputs.Input, dialect: scala.meta.Dialect) = {
+      val parse = implicitly[Parse[Template]]
+      parse(input, dialect)
+    } match {
+      case x: scala.meta.parsers.Parsed.Success[_] => x.tree
+      case x: scala.meta.parsers.Parsed.Error => throw x.details
+    }
+  }
+  object XTensionQuasiquoteMod {
+    private[meta] def parse(input: scala.meta.inputs.Input, dialect: scala.meta.Dialect) = {
+      val parse = implicitly[Parse[Mod]]
+      parse(input, dialect)
+    } match {
+      case x: scala.meta.parsers.Parsed.Success[_] => x.tree
+      case x: scala.meta.parsers.Parsed.Error => throw x.details
+    }
+  }
+  object XTensionQuasiquoteEnumerator {
+    private[meta] def parse(input: scala.meta.inputs.Input, dialect: scala.meta.Dialect) = {
+      val parse = implicitly[Parse[Enumerator]]
+      parse(input, dialect)
+    } match {
+      case x: scala.meta.parsers.Parsed.Success[_] => x.tree
+      case x: scala.meta.parsers.Parsed.Error => throw x.details
+    }
+  }
+  object XTensionQuasiquoteImporter {
+    private[meta] def parse(input: scala.meta.inputs.Input, dialect: scala.meta.Dialect) = {
+      val parse = implicitly[Parse[Importer]]
+      parse(input, dialect)
+    } match {
+      case x: scala.meta.parsers.Parsed.Success[_] => x.tree
+      case x: scala.meta.parsers.Parsed.Error => throw x.details
+    }
+  }
+  object XTensionQuasiquoteImportee {
+    private[meta] def parse(input: scala.meta.inputs.Input, dialect: scala.meta.Dialect) = {
+      val parse = implicitly[Parse[Importee]]
+      parse(input, dialect)
+    } match {
+      case x: scala.meta.parsers.Parsed.Success[_] => x.tree
+      case x: scala.meta.parsers.Parsed.Error => throw x.details
+    }
+  }
+  object XTensionQuasiquoteSource {
+    private[meta] def parse(input: scala.meta.inputs.Input, dialect: scala.meta.Dialect) = {
+      val parse = implicitly[Parse[Source]]
```

Comment:
Since this is all repeated could we not have a method:
```
def parseImpl[T](using parse: Parse[T] )(input: scala.meta.inputs.Input, dialect: scala.meta.Dialect) =
{ 
      parse(input, dialect)
    } match {
      case x: scala.meta.parsers.Parsed.Success[_] => x.tree
      case x: scala.meta.parsers.Parsed.Error => throw x.details
    }
```
and use it everwhere?

Reply:
What do you think about this one? Is it legit to rewrite those into single function?